### ISSUE TITLE
"deployments" could optionally have a name

### ISSUE THREAD

**Issue by rgitzel** [onicagroup/runway#45]
**Background**
Runway projects consist of one or more deployments, each of which contains one or more modules.

Every module has a "name", in that it has a `path` string defined (either provided explicitly, or assumed to be the name of the module's folder), and we can use this in logs:

```
INFO:runway:---- Processing module 'api.sls' for 'dev' in us-east-1 --------------
```

Deployments have nothing to name them, though.  So when we refer to them, we use their index in the list:
```
INFO:runway:Found 1 deployment(s)
INFO:runway:
INFO:runway:======= Processing deployment 0 ===========================
```

**Proposal**
A deployment  will have an optional name which can be used in place of the index for display purposes:

```
INFO:runway:======= Processing deployment `global_resources` ==============
```

The particular name of the value in the YAML file code be... `name`?  `label`?
```
___
deployments:
  - label: global_deployments
    modules:
    ....
```

**Comment by ITProKyle** [comment]
I think `name` would make logical sense. Would be nice to see the option to define deployments as a dict using the `name` as the key as well as supplying the deployments as a list of dicts as shown in your example.

```
___
deployments:
  - name: global_deployments
    modules:
    ....
```
OR
```
___
deployments:
  global_deployments:
    modules:
    ....
```

**Comment by rgitzel** [comment]
I'm now thinking that modules should have explicit (and optional) names, as well.

Presumably you could group your module folders in sub-directories, so a module's path might be `cfn/whatever` and that wouldn't be a great name.

**Comment by rgitzel** [comment]
@ITProKyle Myself, I prefer maps to lists in YAML, so I like the idea.  

But... deployments and modules currently are processed in the order given by the list.  If they are read into a `dict` that ordering would be lost.

**Comment by ITProKyle** [comment]
@rgitzel 
> I'm now thinking that modules should have explicit (and optional) names, as well.
> 
> Presumably you could group your module folders in sub-directories, so a module's path might be `cfn/whatever` and that wouldn't be a great name.

I think this is less of an issue then adding names to a deployment. modules already have `names` since you have to define the path. If the path wouldn't be a good name for the module, they why have the module in that path to begin with?

> But... deployments and modules currently are processed in the order given by the list. If they are read into a `dict` that ordering would be lost.

stacker (a dependency of runway) has logic for handling this that we may be able to reference. the example I provided are both valid formats for defining `stacks` and `*_hooks`.

**Comment by troyready** [comment]
Agreed with the idea of adding the optional `name:` property to the current deployments presuming its needed for #44 and we move forward with it.

> If the path wouldn't be a good name for the module, then why have the module in that path to begin with?
The precedence for this now @ITProKyle is the staticsite modules that allow `name:` to be specified to override the CFN stack names

**Comment by jpoissant** [comment]
I would assume that with the [PR](https://github.com/onicagroup/runway/pull/49) merged this issue can is not _Open_ anymore. Can we update the issue status?

**Comment by GarisonLotus** [comment]
@ITProKyle - I think this issue is actually resolved.  Can we close the issue out?
